src/transform/src/attribute/cardinality.rs

```diff
@@ -0,0 +1,646 @@
+        let mut indexed_columns = BTreeSet::new();
+        for key in keys {
+            if key.len() == 1 {
+                indexed_columns.insert(key[0]);
+            }
+        }
+
+        // worst case scaling factor is 1
+        let mut factor = SymExp::from(1.0);
+
+        for expr in predicates {
+            let predicate_scaling_factor = self.predicate(expr, &indexed_columns);
+
+            // constant scaling factors should be in (0,1]
+            debug_assert!(match predicate_scaling_factor {
+                SymExp::Constant(OrderedFloat(n)) => 0.0 < n && n <= 1.0,
+                _ => true,
+            });
+
+            factor = factor * predicate_scaling_factor;
+        }
+
+        input.clone() * factor
+    }
+
+    fn join(
+        &self,
+        equivalences: &Vec<Vec<MirScalarExpr>>,
+        _implementation: &JoinImplementation,
+        indexed_columns: BTreeMap<usize, usize>,
+        inputs: Vec<&SymExp>,
+    ) -> SymExp {
+        let mut inputs = inputs.into_iter().cloned().collect::<Vec<_>>();
+
+        for equiv in equivalences {
+            // those sources which have a unique key
+            let mut unique_sources = BTreeSet::new();
+            let mut all_unique = true;
+
+            for expr in equiv {
+                if let MirScalarExpr::Column(col) = expr {
+                    if let Some(idx) = indexed_columns.get(col) {
+                        unique_sources.insert(*idx);
+                    } else {
+                        all_unique = false;
+                    }
+                } else {
+                    all_unique = false;
+                }
+            }
+
+            // no unique columns in this equivalence
+            if unique_sources.is_empty() {
+                continue;
+            }
+
+            // ALL unique columns in this equivalence
+            if all_unique {
+                // these inputs have unique keys for _all_ of the equivalence, so they're a bound on how many rows we'll get from those sources
+                // we'll find the leftmost such input and use it to hold the minimum; the other sources we set to 1.0 (so they have no effect)
+                let mut sources = unique_sources.iter();
+
+                let lhs_idx = *sources.next().unwrap();
+                let mut lhs = std::mem::replace(&mut inputs[lhs_idx], SymExp::f64(1.0));
+                for &rhs_idx in sources {
+                    let rhs = std::mem::replace(&mut inputs[rhs_idx], SymExp::f64(1.0));
+                    lhs = SymExp::min(lhs, rhs);
+                }
+
+                inputs[lhs_idx] = lhs;
+
+                // best option! go look at the next equivalence
+                continue;
+            }
+
+            // some unique columns in this equivalence
+            for idx in unique_sources {
+                // when joining R and S on R.x = S.x, if R.x is unique and S.x is not, we're bounded above by the cardinality of S
```

Comment:
After reading Section 2.3 in [today's Reading Group paper](https://www.vldb.org/pvldb/vol9/p204-leis.pdf), I'm thinking that this is not handling a common case from TPC-H: When we have a join between a big table `T` and `Nation`, and there is a filter on the `Nation` side. In this case, the `Nation` side has the unique key, so this code will give the size of `T` as the result. However, the formula in the paper will give the size of `T` scaled by the selectivity of the filter on `Nation`.

Reply:
But unfortunately it's not trivial to implement the formula from the paper, because we would need the domain cardinality of the foreign key, which we currently don't have available.

Reply:
Also, without domain cardinality on the `Nation` side, we also can't really estimate the selectivity of something like `n_name = 'Germany'`.

Reply:
I'm thinking how we could get the domain cardinality, but it's not straightforward. One direction we could explore is to use an already existing index to additionally maintain a the domain cardinality of the indexed key. For the above join, we would advise the user to index `T` on the foreign key, so we would have the relevant index.

The question is how exactly to maintain the domain cardinality based on the index. If we naively do a `Distinct` and then a `count`, then we would maintain a lot of additional memory (proportional to the domain cardinality, which might be a similar size as the original relation in many cases). However, we could create a purpose-built Differential Dataflow operator. (This would be essentially `COUNT DISTINCT`, which might be useful directly for user queries also.) This operator would look into the index at each row change to decide whether to increase/decrease the maintained domain cardinality: when a row is added, it would check whether the key existed in the arrangement before, and if not, then increment the domain cardinality; similarly for row deletions: if the row doesn't exist anymore, then decrease the domain cardinality. I think this wouldn't need any additional memory, just a little bit of extra CPU time when maintaining the index. Unfortunately, creating a new Differential Dataflow operator is kind of a heavy-weight thing from the development effort point of view.